pkgs/by-name/se/sequoia-wot/package.nix

```diff
@@ -0,0 +1,53 @@
+{
+  lib,
+  rustPlatform,
+  fetchFromGitLab,
+  libclang,
+  pkg-config,
+  nettle,
+  openssl,
+  sqlite,
+}:
+rustPlatform.buildRustPackage rec {
+  pname = "sequoia-wot";
+  version = "0.11.0";
+
+  src = fetchFromGitLab {
+    owner = "sequoia-pgp";
+    repo = "sequoia-wot";
+    rev = "v${version}";
+    hash = "sha256-qSf2uESsMGUEvAiRefpwxHKyizbq5Sst3SpjKaMIWTQ=";
+  };
+
+  cargoPatches = [
+    ./01-cargo-lock.patch
+    ./02-cargo-toml.patch
```

Comment:
Why are these needed?

Reply:
Because building the cli requires additional features `sequoia-openpgp/default`, this feature requires a series of packages that do not exist in `Cargo.lock`. Therefore, `Cargo.lock` needs to be patched to introduce additional packages so that nix can add them to the vendor. And if `Cargo.toml` is not patched, it will fail because the package is out of date and needs to be updated.

Frankly, I have no idea what the best practices are for handling this situation, so if you could enlighten me I'd be all ears.

Reply:
> Because building the cli requires additional features `sequoia-openpgp/default`, this feature requires a series of packages that do not exist in `Cargo.lock`. Therefore, `Cargo.lock` needs to be patched to introduce additional packages so that nix can add them to the vendor. And if `Cargo.toml` is not patched, it will fail because the package is out of date and needs to be updated.
> 
> Frankly, I have no idea what the best practices are for handling this situation, so if you could enlighten me I'd be all ears.

@Cryolitia I'm a bit bothered by these, because I worry it will create a maintenance headache in future releases. Could we maybe write an issue to upstream about this? I'm not sure I will be able to do it myself and explain it best for their point of view. Other then that, the PR looks ready. Let's give them a few days to respond and we'll decide afterwards.

Reply:
> @Cryolitia I'm a bit bothered by these, because I worry it will create a maintenance headache in future releases. Could we maybe write an issue to upstream about this? I'm not sure I will be able to do it myself and explain it best for their point of view. Other then that, the PR looks ready. Let's give them a few days to respond and we'll decide afterwards.

I think it's more that a nixpkgs's problem, we should figure out a way to vendor all build-time dependency packages, not only default in `cargo.lock`. I wonder if it not happed before in nixpkgs, I think it should be more common.

Reply:
I have a simple question regarding upstream: If they had included your 2 patches, would they be affected when they'd build their project locally or on their CI? What is the reason they don't enable all features or all the default features of `sequoia-openpgp/default`?

Reply:
> it is true that the default backend should not be set and users should be reminded to make their own choices.

Maybe my problem is that I don't understand this sentence - what is the default backend you are talking about here?

Reply:
You may have noticed that there's one line in `packages.nix`: `buildFeatures = [ "sequoia-openpgp/default" ];`
Without this feature, the build would failed with error below:
```log
sequoia-wot> error: failed to run custom build command for `sequoia-openpgp v1.19.0`
sequoia-wot> Caused by:
sequoia-wot>   process didn't exit successfully: `/build/source/target/release/build/sequoia-openpgp-81b84cfda6136cb6/build-script-build` (exit status: 1)
sequoia-wot>   --- stderr
sequoia-wot>   No cryptographic backend selected.
sequoia-wot>   Sequoia requires a cryptographic backend.  This backend is selected at compile
sequoia-wot>   time using feature flags.
sequoia-wot>   See https://crates.io/crates/sequoia-openpgp#crypto-backends
sequoia-wot> warning: build failed, waiting for other jobs to finish...
```
That means that we are required to choose a cryptographic backend manually and explicitly, they don't provide a defalt backend.

In their [README](https://gitlab.com/sequoia-pgp/sequoia/-/tree/main/openpgp?ref_type=heads#crypto-backends) is documented that, the `openpgp` provides some cryptographic backend, the `default` backend is nettle. Notice that the `default` is just named default, it's not truly the default backend the should be enabled by default. 

It is obvious that, the `sequoia-wot` deliberately set the feature `openpgp/default` backend not enabled by default, but force user —— or package maintainer to choose a backend. So I guess they won't accept the patch.

Reply:
The problem is that, the crates in the patch such as `bzip2` are just required by the feature `sequoia-openpgp/default`, which means a default backend `nettle` named `default`, but may not for other `sequoia-openpgp`'s backend.

Reply:
@Cryolitia thanks for explaining, I pushed a commit that changes this, tell me what do you think about it.

Reply:
Done. You have great insights into Rust, appreciate it.